lib/appsignal/check_in/debouncer.rb

```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+
+module Appsignal
+  module CheckIn
+    class Debouncer
+      SHORT_DEBOUNCE_SECONDS = 0.1
+      LONG_DEBOUNCE_SECONDS = 10
+
+      def initialize
+        # The mutex is used to synchronize access to the events array, the
+        # waker thread and the main thread, as well as queue writes
+        # (which depend on the events array) and closes (so they do not
+        # happen at the same time that an event is added to the debouncer)
+        @mutex = Mutex.new
+        # The thread will be started when an event is first added.
+        @thread = nil
+        @queue = Thread::Queue.new
+        @events = []
+        # The waker thread is used to schedule debounces. It will be started
+        # when an event is first added.
+        @waker = nil
+        # For internal testing purposes.
+        @transmitted = 0
+      end
+
+      def send(event)
+        unless Appsignal.active?
+          Appsignal.internal_logger.debug(
+            "Cannot transmit #{describe([event])}: AppSignal is not active"
+          )
+          return
+        end
+
+        @mutex.synchronize do
+          if @queue.closed?
+            Appsignal.internal_logger.debug(
+              "Cannot transmit #{describe([event])}: AppSignal is stopped"
+            )
+            return
+          end
+          add_event(event)
+          # If we're not already waiting to be awakened from a scheduled
+          # debounce, schedule a short debounce, which will push the events
+          # to the queue and schedule a long debounce.
+          schedule_waker(SHORT_DEBOUNCE_SECONDS) if @waker.nil?
```

Comment:
```suggestion
          schedule_waker(INITIAL_DEBOUNCE_SECONDS) if @waker.nil?
```

Would it be more accurate to say it's the initial sleep/debounce time?

Reply:
Good idea, makes a lot more sense semantically! We should also rename the other one (`LONG_DEBOUNCE_SECONDS`) in turn. How about `FOLLOWING_DEBOUNCE_SECONDS` or `AFTER_TRANSMIT_DEBOUNCE_SECONDS`?

Reply:
We use `ITERATION_IN_SECONDS` in the probes system. Could work here too like `DEBOUNCE_TIME_IN_SECONDS`

Reply:
I went with `BETWEEN_TRANSMISSIONS_DEBOUNCE_SECONDS` 🤷